static/js/composebox_typeahead.js

```diff
@@ -151,38 +152,66 @@ export function should_enter_send(e) {
     return this_enter_sends;
 }
 
+function handle_bulleting_or_numbering($textarea, e) {
+    // handles automatic insertion or removal of bulleting or numbering
+    const before_text = split_at_cursor($textarea.val(), $textarea)[0];
+    const previous_line = list_util.get_last_line(before_text);
+    let to_append = "";
+    // if previous line was bulleted automatically add a bullet to the new line
+    if (list_util.is_bulleted(previous_line)) {
+        // if previous line had only bullet, remove it and stay on the same line
+        if (list_util.strip_bullet(previous_line) === "") {
+            $textarea.range($textarea.caret() - 2, $textarea.caret()).range("");
+            compose_ui.autosize_textarea($textarea);
+            e.preventDefault();
+            return;
+        }
+        // use same bullet syntax as the previous line
+        to_append = previous_line.slice(0, 2);
+    } else if (list_util.is_numbered(previous_line)) {
+        // if previous line was numbered continue numbering with the new line
+        const previous_number_string = previous_line.slice(0, previous_line.indexOf("."));
+        // if previous line had only numbering, remove it and stay on the same line
+        if (list_util.strip_numbering(previous_line) === "") {
+            $textarea
+                .range($textarea.caret() - previous_number_string.length - 2, $textarea.caret())
+                .range("");
+            compose_ui.autosize_textarea($textarea);
+            e.preventDefault();
+            return;
+        }
+        const previous_number = Number.parseInt(previous_number_string, 10);
+        to_append = previous_number + 1 + ". ";
+    }
+    // if previous line was neither numbered nor bulleted, only add
+    // a new line to emulate default behaviour (to_append is blank)
+    // else we add the bulleting / numbering syntax to the new line
+    $textarea.caret("\n" + to_append);
+    compose_ui.autosize_textarea($textarea);
+    e.preventDefault();
+}
+
 export function handle_enter($textarea, e) {
     // Used only if Enter doesn't send.
 
-    // Since this Enter doesn't send, we just want to do
-    // the browser's default behavior for the "Enter" key.
-    // Letting the browser handle it works great if the
-    // key actually pressed was Enter or Shift-Enter.
-
-    // But the default browser behavior for Ctrl/Alt/Meta
-    // + Enter is to do nothing, so we need to emulate
-    // the browser behavior for "Enter" in those cases.
-    //
-    // We do this using caret and range from jquery-caret.
-    const has_non_shift_modifier_key = e.ctrlKey || e.metaKey || e.altKey;
-    if (!has_non_shift_modifier_key) {
-        // Use the native browser behavior.
-        return;
-    }
-
```

Comment:
Can you comment on why deleting this will have no regressions?

Reply:
Now that we need to check the state of the previous line (if it's bulleted / numbered and take action accordingly) every time enter (or something to the same effect) is used, there are no cases where we just let the native behaviour take over, as we emulate that as part of the `handle_bulleting_or_numbering` function (even in cases where bulleting / numbering isn't needed). Hence, not using the native browser behaviour simplifies the code

Reply:
I guess we will need to look for regressions in testing. I couldn't find any when I did.

Reply:
Yeah, I can confirm that this should have no functional impact as long as we implement the emulation correctly, which we need to do. Could likely use a pass on comments though -- I'll push some changes to them.